Fixes to the guard AI

GuybrushGit opened the pull request:
- Fixed attacking PvP flagged mobs not flagging the player for contested PvP (https://github.com/elysium-project/server/issues/2094).
- Fixed guards not protecting nearby friendly NPCs.

BrynjaDarrow commented:
Wasn't the first part already fixed with https://github.com/elysium-project/server/pull/588?
"Attacking or healing PvP flagged NPCs not flagging the caster"

GuybrushGit commented:
@BrynjaDarrow no its a different flag.

GuybrushGit commented:
Removed the AI changes they did not work if the player never moved.
Replaced with a faction check that does the same without adding a bunch of code to the AI. 
There is no flag or identifier for contested factions so I had to hardcode the ids.

brotalnia commented:
Actually you can check the faction template flags instead of hardcoding the faction ids.

FACTION_TEMPLATE_FLAG_CONTESTED_GUARD = 0x00001000 (4096)

Here is a list of faction templates with it:
> Template 121, Faction 21, Flags 6177, FriendlyMask 0, HostileMask 8
> Template 475, Faction 369, Flags 6177, FriendlyMask 0, HostileMask 8
> Template 637, Faction 470, Flags 6177, FriendlyMask 0, HostileMask 8
> Template 854, Faction 577, Flags 6177, FriendlyMask 0, HostileMask 8
> Template 1254, Faction 609, Flags 4129, FriendlyMask 0, HostileMask 8
> Template 1615, Faction 169, Flags 6177, FriendlyMask 0, HostileMask 8
> Template 1625, Faction 529, Flags 4129, FriendlyMask 0, HostileMask 8
> Template 1635, Faction 169, Flags 6161, FriendlyMask 0, HostileMask 0

As you can see those are all the factions you listed.

GuybrushGit commented:
Those are for the guards, we want the units protected by the guards. For example for BB its 120 not 121.

GuybrushGit commented:
But then I'd need to iterate through all the faction templates just to match factions that equal the faction of a contested guard. 
This function gets called every time someone deals damage to a unit, it seems a waste of CPU to do that when the factions are constant.
